[SPARK-29093][PYTHON][ML] Remove automatically generated param setters in _shared_params_code_gen.py

huaxingao opened the pull request:
### What changes were proposed in this pull request?
Remove automatically generated param setters in _shared_params_code_gen.py


### Why are the changes needed?
To keep parity between scala and python


### Does this PR introduce any user-facing change?
Yes
Add some setters in Python ML XXXModels 


### How was this patch tested?
unit tests

huaxingao committed:
[SPARK-29093][PYTHON][ML] Remove automatically generated param setters in _shared_params_code_gen.py

SparkQA commented:
**[Test build #112559 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112559/testReport)** for PR 26232 at commit [`a67163e`](https://github.com/apache/spark/commit/a67163e285876651dbb7eeb1d0ee66aa30dbce31).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17524/
Test PASSed.

SparkQA commented:
**[Test build #112559 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112559/testReport)** for PR 26232 at commit [`a67163e`](https://github.com/apache/spark/commit/a67163e285876651dbb7eeb1d0ee66aa30dbce31).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112559/
Test FAILed.

huaxingao committed:
add more doc tests

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17536/
Test PASSed.

SparkQA commented:
**[Test build #112571 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112571/testReport)** for PR 26232 at commit [`d076f9e`](https://github.com/apache/spark/commit/d076f9e4515ea4c52e760ceb8a41c645dd2cc65d).

SparkQA commented:
**[Test build #112571 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112571/testReport)** for PR 26232 at commit [`d076f9e`](https://github.com/apache/spark/commit/d076f9e4515ea4c52e760ceb8a41c645dd2cc65d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DCT(JavaUnaryTransformer, JavaMLReadable, JavaMLWritable):`
  * `class ElementwiseProduct(JavaUnaryTransformer, JavaMLReadable, JavaMLWritable):`
  * `class NGram(JavaUnaryTransformer, JavaMLReadable, JavaMLWritable):`
  * `class Normalizer(JavaUnaryTransformer, JavaMLReadable, JavaMLWritable):`
  * `class PolynomialExpansion(JavaUnaryTransformer, JavaMLReadable, JavaMLWritable):`
  * `class RegexTokenizer(JavaUnaryTransformer, JavaMLReadable, JavaMLWritable):`
  * `class Tokenizer(JavaUnaryTransformer, JavaMLReadable, JavaMLWritable):`
  * `class JavaUnaryTransformer(JavaParams, UnaryTransformer):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112571/
Test FAILed.

huaxingao committed:
not using UnaryTransformer for now

SparkQA commented:
**[Test build #112583 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112583/testReport)** for PR 26232 at commit [`bff5a70`](https://github.com/apache/spark/commit/bff5a70e2bdcd23fccdb03647fd70f0197440cac).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17546/
Test PASSed.

SparkQA commented:
**[Test build #112583 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112583/testReport)** for PR 26232 at commit [`bff5a70`](https://github.com/apache/spark/commit/bff5a70e2bdcd23fccdb03647fd70f0197440cac).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DCT(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class ElementwiseProduct(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable,`
  * `class NGram(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class Normalizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class PolynomialExpansion(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable,`
  * `class RegexTokenizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`
  * `class Tokenizer(JavaTransformer, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112583/
Test PASSed.

zhengruifeng commented:
1, why you remove the `__metaclass__ = ABCMeta` in wrapper.py?
2, can we add the missing since annotations for new added setters?
I can not 'add single comment' or 'start a review', I guess that is beacuse this PR is still to large although we had finished some preposed PRs.

huaxingao commented:
@zhengruifeng 
Sorry, I accidentally removed  ```__metaclass__ = ABCMeta``` . Will add back.

For the newly added setters in Transformers, most of them are new in 3.0.0. I will double check to make sure all the newly added one have ```@since("3.0.0")```. For the newly added setters in Estimators, all of them are existing ones. I initially tried to trace back the first version these setters appeared, and added a bunch of since annotations. Then I checked the documentation,  and found out only the parameters that are specific to the Estimators/Transformers have since annotations in the setters, the parameters in HasXXX don't have since annotations in the setters. Here is one example for ```PCA```:

![image](https://user-images.githubusercontent.com/13592258/67495764-42238d00-f630-11e9-97f1-26395369dcab.png)

I guess I will only add ```@since("3.0.0")``` for the setters that are really new in 3.0.0. For the newly added setters in Estimators, since they are actually preexisting ones, I will not add since annotation. I guess I will remove the ones I have added. Does this sound good to you?

huaxingao committed:
address comments

SparkQA commented:
**[Test build #112610 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112610/testReport)** for PR 26232 at commit [`03c2e4a`](https://github.com/apache/spark/commit/03c2e4a6c7581402de0c1067c31f5655d9992364).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17575/
Test PASSed.

SparkQA commented:
**[Test build #112610 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112610/testReport)** for PR 26232 at commit [`03c2e4a`](https://github.com/apache/spark/commit/03c2e4a6c7581402de0c1067c31f5655d9992364).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112610/
Test PASSed.

zhengruifeng commented:
merged to master, thanks @huaxingao !
After this PR, the py hierarchy should be highly similar to the scala side.

huaxingao commented:
Thanks a lot! @zhengruifeng

zero323 commented on python/pyspark/ml/clustering.py:
```diff
@@ -873,6 +1049,31 @@ class LDAModel(JavaModel, _LDAParams):
+        """
+        Sets the value of :py:attr:`topicDistributionCol`.
+
+        >>> algo = LDA().setTopicDistributionCol("topicDistributionCol")
```
If anything, shouldn't we have model specific dosctring here?  This one just duplicates what we already have in estimator:

https://github.com/apache/spark/blob/b389b8c5f0650a7e63098f18437fcaa29998732a/python/pyspark/ml/clustering.py#L1159-L1160

huaxingao replied:
I am in the middle of rechecking this PR and other ml parity PRs. I will have a follow up PR soon to fix this along with a few other problems I found.

zero323 replied:
Thanks @huaxingao. If you don't mind me asking ‒ should I expect any serious shifts?

huaxingao replied:
No. Just fix the small things I overlooked before, for example, I have setters in ```TrainValidationSplitModel/CrossValidatorModel``` because these setters are used  in ```_from_java```.  I will remove the setters and use _set instead.

zero323 replied:
Thanks @huaxingao.

zero323 commented on python/pyspark/ml/feature.py:
```diff
@@ -707,6 +855,34 @@ class CountVectorizerModel(JavaModel, _CountVectorizerParams, JavaMLReadable, Ja
+        return self._set(outputCol=value)
+
+    @since("3.0.0")
+    def setMinTF(self, value):
```
This one is already present in 2.4  ([L916-L921](https://github.com/huaxingao/spark/blob/03c2e4a6c7581402de0c1067c31f5655d9992364/python/pyspark/ml/feature.py#L916-L921))

zero323 commented on python/pyspark/ml/feature.py:
```diff
@@ -707,6 +855,34 @@ class CountVectorizerModel(JavaModel, _CountVectorizerParams, JavaMLReadable, Ja
+        return self._set(minTF=value)
+
+    @since("3.0.0")
+    def setBinary(self, value):
```
Ditto, [L923-L928](https://github.com/huaxingao/spark/blob/03c2e4a6c7581402de0c1067c31f5655d9992364/python/pyspark/ml/feature.py#L923-L928)

huaxingao replied:
The fix were in a few hours ago. Please let me know if you see other problems. Thanks! @zero323
